src/NuGet.Core/NuGet.Protocol/Utility/Utf8JsonStreamReader.cs

```diff
@@ -0,0 +1,175 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Buffers;
+using System.IO;
+using System.Linq;
+using System.Text.Json;
+
+namespace NuGet.Protocol
+{
+    /// <summary>
+    /// This struct is used to read over a memory stream in parts, in order to avoid reading the entire stream into memory.
+    /// It functions as a wrapper around <see cref="Utf8JsonStreamReader"/>, while maintaining a stream and a buffer to read from.
+    /// </summary>
+    internal ref struct Utf8JsonStreamReader
```

Comment:
>internal ref struct Utf8JsonStreamReader [](http://example.com/codeflow?start=4&length=40)

I didn't like duplicating this class, but I wasn't sure how to get access to it otherwise.